core/Tracker/Model.php

```diff
@@ -398,10 +398,20 @@ public function findVisitor($idSite, $configId, $idVisitor, $fieldsToRead, $shou
         } elseif ($shouldMatchOneFieldOnly) {
             $visitRow = $this->findVisitorByConfigId($configId, $select, $from, $configIdWhere, $configIdbindSql);
         } else {
-            $visitRow = $this->findVisitorByVisitorId($idVisitor, $select, $from, $visitorIdWhere, $visitorIdbindSql);
+            if (!empty($idVisitor)) {
+                $visitRow = $this->findVisitorByVisitorId($idVisitor, $select, $from, $visitorIdWhere, $visitorIdbindSql);
+            } else {
+                $visitRow = false;
+            }
 
             if (empty($visitRow)) {
-                $configIdWhere .= ' AND user_id IS NULL ';
+                $configIdWhere .= ' AND ( user_id IS NULL ';
```

Comment:
@diosmosis @mattab noticing this would create a new visit when userId was previously set, and then user logs out using resetUserId. I suppose this is not what we want as we also don't create a new visit when userId is set initially. I suppose we maybe really only always want to add ` AND ( user_id IS NULL OR user_id = ? )` when userId is not empty?

Reply:
Also when user is found by idVisitor, then we would also not create a new visit after resetUserId so we need to make sure to have consistent behavior... I will later change the query I reckon to only compare userId when a userId is set

Reply:
If userId is null, we'd want to look for ( user_id IS NULL ) though, correct? Otherwise it would match the logged out user. Then again, ( user_id IS NULL OR user_id = ? ) would also match the user after they login (ie, logged out => login), so maybe both behaviors are desired?

Reply:
That's current desired/ implemented behaviour AFAIK and that's how it also works when it matches the visit by visitorId.  See thread in https://github.com/matomo-org/developer-documentation/pull/320

We just need to make sure to always have consistent behaviour on log in and log out. Either we always create a new visit on log in and log out, or never.

Reply:
I thought the point of separating user ID from visitor ID was so you could tell when a user logged in/out. Ie, it would count as the same visitor, but could as different users?

Reply:
I'm not sure? If that's the case, then I think this wasn't implemented.

Reply:
I think you're right it was never implemented. It does seem more useful to me though. I guess it could go in another issue.